src/arp.rs

```diff
@@ -70,51 +76,55 @@ pub struct ArpPacket {
     pub dest_addr: Ipv4Addr,
 }
 
-named!(hw_addr_type<&[u8], HardwareAddressType>, map_opt!(u16!(Big), to_hw_addr_type));
-named!(proto_addr_type<&[u8], ProtocolAddressType>, map_opt!(u16!(Big), to_proto_addr_type));
-named!(operation<&[u8], Operation>, map_opt!(u16!(Big), to_operation));
-named!(mac_address<&[u8], MacAddress>, map!(take!(6), to_mac_address));
-named!(address<&[u8], Ipv4Addr>, map!(take!(4), to_ipv4_address));
-named!(arp_packet<&[u8], ArpPacket>, do_parse!(
-    hw_addr_type: hw_addr_type >>
-    proto_addr_type: proto_addr_type >>
-
-    hw_addr_size:  be_u8 >>
-    proto_addr_size: be_u8 >>
-
-    operation: operation >>
-
-    src_mac: mac_address >>
-    src_addr: address >>
+fn parse_hw_addr_type(input: &[u8]) -> IResult<&[u8], HardwareAddressType> {
+    let (input, hw_addr_type) = number::streaming::be_u16(input)?;
 
-    dest_mac: mac_address >>
-    dest_addr: address >>
-
-    ({ ArpPacket{
-        hw_addr_type: hw_addr_type,
-        proto_addr_type: proto_addr_type,
+    Ok((input, hw_addr_type.into()))
+}
 
-        hw_addr_size: hw_addr_size,
-        proto_addr_size: proto_addr_size,
+fn parse_proto_addr_type(input: &[u8]) -> IResult<&[u8], ProtocolAddressType> {
+    let (input, proto_addr_type) = number::streaming::be_u16(input)?;
 
-        operation: operation,
+    Ok((input, proto_addr_type.into()))
+}
 
-        src_mac: src_mac,
-        src_addr: src_addr,
+fn parse_operation(input: &[u8]) -> IResult<&[u8], Operation> {
+    let (input, operation) = number::streaming::be_u16(input)?;
 
-        dest_mac: dest_mac,
-        dest_addr: dest_addr,
-    }})
-));
+    Ok((input, operation.into()))
+}
 
-pub fn parse_arp_pkt(i: &[u8]) -> IResult<&[u8], ArpPacket> {
-    arp_packet(i)
+pub fn parse_arp_pkt(input: &[u8]) -> IResult<&[u8], ArpPacket> {
+    let (input, hw_addr_type) = parse_hw_addr_type(input)?;
+    let (input, proto_addr_type) = parse_proto_addr_type(input)?;
+    let (input, hw_addr_size) = number::streaming::be_u8(input)?;
+    let (input, proto_addr_size) = number::streaming::be_u8(input)?;
+    let (input, operation) = parse_operation(input)?;
+    let (input, src_mac) = ethernet::mac_address(input)?;
+    let (input, src_addr) = ipv4::address(input)?;
+    let (input, dest_mac) = ethernet::mac_address(input)?;
+    let (input, dest_addr) = ipv4::address(input)?;
+
+    Ok((
+        input,
+        ArpPacket {
+            hw_addr_type,
+            proto_addr_type,
+            hw_addr_size,
+            proto_addr_size,
+            operation,
+            src_mac,
+            src_addr,
+            dest_mac,
+            dest_addr,
+        },
+    ))
 }
 
 #[cfg(test)]
 mod tests {
     use super::{
-        arp_packet, ArpPacket, HardwareAddressType, MacAddress, Operation, ProtocolAddressType,
```

Comment:
`arp_packet` is not public, I remove it because `parse_arp_pkt` was just a wrapper.